lib/auth/accountrecovery_test.go

```diff
@@ -517,6 +517,115 @@ func TestApproveAccountRecovery_WithLock(t *testing.T) {
+	})
+	require.NoError(t, err)
+
+	// Test locks are removed.
+	user, err := srv.Auth().GetUser(u.username, false)
+	require.NoError(t, err)
+	require.False(t, user.GetStatus().IsLocked)
+	require.True(t, user.GetStatus().RecoveryAttemptLockExpires.IsZero())
+	require.True(t, user.GetStatus().LockExpires.IsZero())
+
+	// Test login attempts are removed.
+	attempts, err := srv.Auth().GetUserLoginAttempts(u.username)
+	require.NoError(t, err)
+	require.Len(t, attempts, 0)
+
+	// Test adding MFA devices.
+	approvedToken, err = srv.Auth().createRecoveryToken(ctx, u.username, UserTokenTypeRecoveryApproved, types.UserTokenUsage_USER_TOKEN_RECOVER_MFA)
+	require.NoError(t, err)
+
+	cases := []struct {
+		name       string
+		getRequest func() *proto.CompleteAccountRecoveryRequest
+	}{
+		{
+			name: "add new TOTP device",
+			getRequest: func() *proto.CompleteAccountRecoveryRequest {
+				secrets, err := srv.Auth().RotateUserTokenSecrets(context.TODO(), approvedToken.GetName())
+				require.NoError(t, err)
+
+				otpCode, err := totp.GenerateCode(secrets.GetOTPKey(), srv.Clock().Now())
+				require.NoError(t, err)
+
+				return &proto.CompleteAccountRecoveryRequest{
+					NewDeviceName:           "new-otp",
+					RecoveryApprovedTokenID: approvedToken.GetName(),
+					NewAuthnCred: &proto.CompleteAccountRecoveryRequest_NewMFAResponse{NewMFAResponse: &proto.MFARegisterResponse{
+						Response: &proto.MFARegisterResponse_TOTP{TOTP: &proto.TOTPRegisterResponse{Code: otpCode}},
+					}},
+				}
+			},
+		},
+		{
+			name: "add new U2F device",
+			getRequest: func() *proto.CompleteAccountRecoveryRequest {
+				u2fRegResp, _, err := getMockedU2FAndRegisterRes(srv, approvedToken.GetName())
+				require.NoError(t, err)
+
+				return &proto.CompleteAccountRecoveryRequest{
+					NewDeviceName:           "new-u2f",
+					RecoveryApprovedTokenID: approvedToken.GetName(),
+					NewAuthnCred: &proto.CompleteAccountRecoveryRequest_NewMFAResponse{NewMFAResponse: &proto.MFARegisterResponse{
+						Response: &proto.MFARegisterResponse_U2F{U2F: u2fRegResp},
+					}},
+				}
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			// Get proto request.
+			req := c.getRequest()
+
+			// Change authentication.
+			err = srv.Auth().CompleteAccountRecovery(ctx, c.getRequest())
+			require.NoError(t, err)
+
+			// Test events emitted.
+			event := mockEmitter.LastEvent()
+			require.Equal(t, event.GetType(), events.MFADeviceAddEvent)
+			require.Equal(t, event.GetCode(), events.MFADeviceAddEventCode)
+			require.Equal(t, event.(*apievents.MFADeviceAdd).UserMetadata.User, u.username)
+
+			// Test new device has been added.
+			mfas, err := srv.Auth().Identity.GetMFADevices(ctx, u.username, false)
+			require.NoError(t, err)
+
+			deviceNames := make([]string, 0, len(mfas))
+			for _, mfa := range mfas {
+				deviceNames = append(deviceNames, mfa.GetName())
+			}
+			require.Contains(t, deviceNames, req.NewDeviceName)
```

Comment:
nit: doesn't matter a whole lot, but:

```suggestion
			found := false
			for _, mfa := range mfas {
				if mfa.GetName() == req.NewDeviceName {
					found = true
					break
				}
			}
			require.True(t, found, "MFA device %q not found", req.NewDeviceName)
```